src/bootchooser.c

```diff
@@ -284,6 +286,57 @@ static gboolean grub_set_primary(RaucSlot *slot) {
 	return res;
 }
 
+static gboolean uboot_env_get(const gchar *key, GString **value) {
```

Comment:
Might be a bit inconsistent here to have `GString **` in get and `gchar *` in set function. As we actually pass a GString to `uboot_env_set()`, anyway, it might be the simplest solution to change its arg to `GString *`.

Reply:
Are we sure we want to change the `value` argument to `GString *`? Doesn't that require instantiation and freeing of temporary `GString` objects for every place where `uboot_env_set()` is called with a string literal as value? (e.g. `res = uboot_env_set(key, "3");` in line 421.)

To gain consistency between `uboot_env_{set,get}()` I can change the value argument of `uboot_env_get()` to `char *` and add a length argument, e.g `static gboolean uboot_env_get(const char *key, char *value, size_t len)` and return FALSE is the value buffer is too small to hold the env-string?

Reply:
Ping ?

Reply:
Sorry for the delay. I wanted to test this a bit on hardware but do not have a board with u-boot present atm. Maybe I have to drop the hardware test and trust code reading and testing. But I would like to have @jluebbe's eyes on that before merging, as this touches critical infrastructure.

Well, calling `uboot_env_set` with fixed strings is more ore less a simplification that I am not yet really happy with, yet. Maybe one does not want 3 but only 2 or only 1 attempt for newly activated slots. Thus in the future this could become configurable.
Concerning your remarks for the return type of uboot_env_get I would not like to instancaite and pass buffers from outside. We could do the same as with the `GString`, returning a newly allocated `gchar*`. This is what all other functions do, too (sometimes as return value, sometimes as point-pointer argument).
But in general, I would not see this discussion as a blocker. There is no public API designed, thus we are free to refactor the code and methods if that should be required.